### ISSUE TITLE
Idea for implementing concurrent functions

### ISSUE THREAD

**Issue by maxnordlund** [fish-shell/fish-shell#2728]
I'm been thinking about this for some time, and think I've come up with something that can be implemented without to much trouble. But also allow to reuse most of the non-thread safe code.

## The Idea
The basic idea is to have `Services` (or `Deamons`) classes that enforce the boundary between threads. 

### What do I mean with that?
Well I was thinking having a small set of superclasses which implement the necessary infrastructure for concurrent access and exposing this as static methods.

So instead of creating a new instance, or having a singleton, you call `MyClass.get` which will return an instance for you to use. Then you use it until your done, and return it back. This could/should probably use a destructor to make a bit easier. I would like to return a [`Future`][] to allow for very natural use, then just remember to call `wait`, and with a custom subclass this could also return it to the service.

### Why return the instance?
To allow use of thread/process/singelton state/caches/... pools. This is also why I think it will be a very small set of superclasses, one for a thread, one for global state (singleton), one for processes and one that caches the result.

## Common flow using this
The shell, a global service, gets input from the user, and sends it to the parser, a thread pool. Then the resulting AST is inspect to see if there are errors to be displayed to the user, syntax highlighting and autosuggestion.

If the user asks for autocompletion (or maybe also the autosuggestion), the autocomplete service is called. This would need to be a singleton broker to allow caching, or just a thread pool. It will call a process service to run the autocomplete command in the background, which in turn also calls the parser service. Then it wait for two types of input, search terms from the user and valid entries from the command. It will then do its thing and send the result back to the screen service to actually be rendered. (This might be the shell, not sure)

The `jobs` control would use another process service to make it easy to query and leak autocomplete commands to the user. However the process service could support pipelines directly to make it easier to use, and also understanding that fish functions should be run using the interpreter service.

---

### Caching autocomplete
Its probably only worth it for large inputs, like `git log`, or very slow autocomplete commands, like `rake -T` or `find . <something very deep>`. To make it work the autocomplete author needs to supply two more functions/hooks.

The first to supply a hash key, in in case of `git` this would be the closest git root directory. This means even if you `cd` around it can reuse the cache.

The second to supply an invalidation key, which again in `git`s case would be the HEAD commit sha. This could also be the last modification date or something completely different.

Why two separate functions/hooks? See the `git` case for example of why it would be very useful.

---

I most likely missed a bunch of stuff, but at least this may be used as a starting point to get things rolling. I also think having an documented architecture makes it easier to get stuff done :tm: But more importantly allow people like me to jump in end help. If this, whatever it may become, is decided to be the way forward, I would really like it to become a proper document, on the wiki or as a file in the repo.

If an owner would like to add/change this please do so. I want it to be a living document. What say ye?

[`Future`]: http://en.cppreference.com/w/cpp/thread/future

**Comment by pickfire** [comment]
I like this idea, I like to do things concurrently.

**Comment by ridiculousfish** [comment]
Thanks for posting this! By "process service," do you mean that it maintains a separate process, as in OS-level process (i.e. fork)?

**Comment by maxnordlund** [comment]
Yes, since that is needed to actually the programs in a pipeline, for example in `grep foo bar.md | wc -l` both `grep` and `wc` needs their own process and have their standard in/out linked. But I think the process service should take care of that, to allow the outside to view it as a single process with a single set of stdin/out/err.

**Comment by ridiculousfish** [comment]
That sort of pipeline happens to be easy, because there's no fish code involved in the pipeline. Just fork and exec, or posix_spawn. Though I'm not following what you mean by the outside viewing it as a single process?

The main difficulty with concurrency in shells is that you can use multiple threads or multiple processes, but not both. fish has some very lightweight tasks, like syntax highlighting, where using multiple processes would be prohibitively expensive (well, at least on OS X). That leaves threads. But now if we want background-able fish code, we need to use threads to emulate multiple processes, which is annoying because there's a lot of per-process global state (cwd, pid, etc) that we would have to instance. We can't have two threads trying to set the cwd at the same time, they'll step on each others' toes.

Anyways, assuming we want to keep using threads, I think that would be a good first step: Identify all the places that fish uses per-process global state, like the cwd or foreground process group, and then instance that state into a struct that we pass around to all of the functions that need it  - basically isolating the places where we use global state.

**Comment by maxnordlund** [comment]
What I would like to see is that the internal structure representing a OS process, with stdin/out/err looks on the outside the same as the internal structure representing a pipeline of those and the same for fish functions being interpreted.

In my model, each service instance would essentially own the thread they are currently executing in, and if they need to keep "global" state, like cwd and local variables, they can do so just by having it has instance variables. The thread safety is instead enforced at the service class level, using static methods to expose the needed interface. This probably boils down to something as simple as `Foo.get(param, ...)`.

For external processes, I think it would be easiest if they have a corresponding thread in fish to talk to, unless writing to stdin and reading from stdout/err can be easily implemented without them.

There will of course be some sort of global state, like the environment, and non-local variables. For those I suggested having a service that only has a single instance, effectively acting like a singleton. On the other hand, for that it might be better to just use thread safe container from boost.

In either case, every time a something, e.g. fish functions, need to access some resources they use a service. This could be talking to a process, using ncurses to paint something, or something more internal like the global state.

[![fish shell services idea](https://cloud.githubusercontent.com/assets/664779/13033482/265694c6-d318-11e5-88b7-8f51e0528913.png)](https://docs.google.com/drawings/d/1dM21sr1UfJIryDWEEEFQWGHjN1PRwKaJcV7npSqvty8/edit?usp=sharing)
